src/functional/fencil_processors/processor_interface.py

```diff
@@ -0,0 +1,155 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+"""
+Interface for fencil processors.
+
+Fencil processors are functions which take an ``iterator.ir.FencilDefinition`` along with input values for the fencil.
+Those which execute the fencil with the given arguments (possibly by generating code along the way) are fencil executors.
+Those which generate / any kind of string based on the fencil and (optionally) input values are fencil formatters.
+
+For more information refer to ``gt4py/docs/functional/architecture/007-Fencil-Processors.md``
+"""
+from __future__ import annotations
+
+from functools import update_wrapper
+from typing import Protocol, TypeVar, runtime_checkable
+
+from functional.iterator.ir import FencilDefinition
+
+from .source_modules import SourceModule
+
+
+PROCESSOR_RETURN_T = TypeVar("PROCESSOR_RETURN_T", covariant=True)
+PROCESSOR_KIND_T = TypeVar("PROCESSOR_KIND_T", bound="FencilProcessorProtocol", covariant=True)
+
+
+class FencilProcessorFunction(Protocol[PROCESSOR_RETURN_T]):
+    def __call__(self, fencil: FencilDefinition, *args, **kwargs) -> PROCESSOR_RETURN_T:
+        ...
+
+
+@runtime_checkable
+class FencilProcessorProtocol(
+    FencilProcessorFunction[PROCESSOR_RETURN_T], Protocol[PROCESSOR_RETURN_T, PROCESSOR_KIND_T]
+):
+    __name__: str = ""  # dummy to convince mypy that a __name__ is present for error messages
+
+    @classmethod
+    def kind(cls) -> type[PROCESSOR_KIND_T]:
+        ...
+
+
+class FencilFormatter(FencilProcessorProtocol[str, "FencilFormatter"], Protocol):
+    @classmethod
+    def kind(cls) -> type[FencilFormatter]:
+        return FencilFormatter
+
+
+def fencil_formatter(
+    func: FencilProcessorFunction[str],
+) -> FencilProcessorProtocol[str, FencilFormatter]:
+    """
+    Wrap a formatter function in a ``FencilFormatter`` instance.
+
+    Examples:
+    ---------
+    >>> @fencil_formatter
+    ... def format_foo(fencil: FencilDefinition, *args, **kwargs) -> str:
+    ...     '''A very useless fencil formatter.'''
+    ...     return "foo"
+
+    >>> ensure_processor_kind(format_foo, FencilFormatter)
+    """
+
+    class _FormatterClass(FencilFormatter):
+        def __call__(self, fencil: FencilDefinition, *args, **kwargs) -> str:
+            return func(fencil, *args, **kwargs)
+
+    formatter_instance = _FormatterClass()
+    update_wrapper(formatter_instance, func)
+    return formatter_instance
```

Comment:
I will try that. Funnily this is almost exactly what my first draft looked like weeks ago.

Reply:
And the reason why I abandoned it is that mypy righly points out that `func` does not have an attribute `kind`. I see multiple ways to circumvent that problem but none of them seem very good.

Reply:
It doesn't work without some type ignores, however, because it is not necessary anymore to wrap the function into a class, the reason for using `__name__` in the error message goes away, and therefore the need to type declare `__name__` on the protocol class etc.

Reply:
:+1: